### PR TITLE
plat-stm32mp1: bumping to pinctrl framework 

### PR DESCRIPTION
**Pull request by etienne-lms**
Bump stm32_gpio driver to recently merged  pinctrl device driver probing support.

For STM32MP13 variant to comply with pinctrl support (and not panic on driver probe errors), I needed to update STM32MP13 DTS and DTSI file and 2 stm32 drivers (stm32_etzpc and stm32_rstctrl). They are part of this series.

A patch fixing a pinctrl driver probe function in under review at https://github.com/OP-TEE/optee_os/pull/6081.

### CODE REVIEW

**Comment by etienne-lms** [comment]
This series it too long. I'll shorten it and rebase it.

File: core/drivers/stm32_gpio.c [hunk 1]
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
...
+	assert(fdt && pinctrl_node);
+
+	fdt_for_each_subnode(pinmux_node, fdt, pinctrl_node) {
+		if (fdt_getprop(fdt, pinmux_node, "pinmux", &count))
+			pin_count += (size_t)count / sizeof(uint32_t);
+		else if (count != -FDT_ERR_NOTFOUND)
+			panic();
+	}
+
+	loc_conf = calloc(1, sizeof(*loc_conf) + sizeof(*pinctrl) * pin_count);
```
Comment (GseoC) [inline core/drivers/stm32_gpio.c:780]: Is is better to do this shortcut or explicitely allocate memory for pinctrl structures in a separate calloc?

File: core/drivers/stm32_gpio.c [hunk 2]
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
...
+	pinconf->ops = &stm32_pinctrl_ops;
+	pinconf->priv = &loc_conf->array_ref;
+
+	loc_conf->array_ref.count = pin_count;
+	pinctrl = loc_conf->array_ref.pinctrl;
+
+	count = 0;
+	fdt_for_each_subnode(pinmux_node, fdt, pinctrl_node) {
+		int found = 0;
+
```
Comment (GseoC) [inline core/drivers/stm32_gpio.c:796]: Can you detail the error case please ?

File: core/drivers/stm32_gpio.c [hunk 3]
```diff
@@ -51,6 +52,16 @@
 
 #define DT_GPIO_BANK_NAME0	"GPIOA"
 
+/*
+ * struct stm32_pinctrl_array - Array of pins in a pin control state
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:56]: A "pin control state" here relates to `struct pinctrl_state`: it is a configuration set for a set of pins for a specific pin muxing use. For example a min muxing state to I2C bus wires when I2C bus is active, and other set when I2C bus is sleeping.

File: core/drivers/stm32_gpio.c [hunk 4]
```diff
@@ -51,6 +52,16 @@
...
+ * @count: Number of cells in @pinctrl
+ * @pinctrl: Pin controls configuration
+ */
+struct stm32_pinctrl_array {
+	size_t count;
+	struct stm32_pinctrl pinctrl[];
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:62]: no: field `pinctrl` is not a pointer to an array allocated somewhere, it is the start point of the array.

File: core/drivers/stm32_gpio.c [hunk 5]
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
...
 		/* Check GPIO bank clock/base address against platform */
 		ckeck_gpio_bank(fdt, bank, pinctrl_node);
+#endif
 
 		if (found < count) {
 			struct stm32_pinctrl *ref = &pinctrl[found];
 
 			ref->bank = (uint8_t)bank;
 			ref->pin = (uint8_t)pin;
+#ifdef CFG_DRIVERS_PINCTRL
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:351]: These switch will be shortly removed (once the [full patches series](https://github.com/OP-TEE/optee_os/pull/6050) is merged) as I expect stm32_gpio to mandate `CFG_DRIVERS_PINCTRL=y`.

File: core/drivers/stm32_gpio.c [hunk 6]
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
...
+			ref->cfg.mode = mode;
+			ref->cfg.otype = opendrain ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:353]: i'll check that. maybe in a later P-R, to not mix a fix and this change taht prepares for pinctrl support.

File: core/drivers/stm32_gpio.c [hunk 7]
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
 	clk_disable(bank->clock);
 }
 
+#ifdef CFG_DRIVERS_PINCTRL
+static TEE_Result stm32_pinctrl_conf_apply(struct pinconf *conf)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:667]: this function must comply with `struct pinctrl_ops::pinctrl_apply`

File: core/drivers/stm32_gpio.c (hunk 1 above)
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:780]: I think it is preferred, to lower the heap fragmentaiton.

File: core/drivers/stm32_gpio.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:796]: you mean why do we panic at line 800?

Comment (GseoC) [reply core/drivers/stm32_gpio.c:796]: Yes indeed :)

File: core/drivers/stm32_gpio.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_gpio.c:351]: Guess that's fine then

File: core/drivers/stm32_gpio.c [hunk 8]
```diff
@@ -51,6 +52,16 @@
 
 #define DT_GPIO_BANK_NAME0	"GPIOA"
 
+/*
+ * struct stm32_pinctrl_array - Array of pins in a pin control state
+ * @count: Number of cells in @pinctrl
+ * @pinctrl: Pin controls configuration
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:58]: I went for `Pin control configuration`

File: core/drivers/stm32_gpio.c [hunk 9]
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
...
+#ifdef CFG_DRIVERS_PINCTRL
+			ref->cfg.mode = mode;
+			ref->cfg.otype = opendrain ? 1 : 0;
+			ref->cfg.ospeed = speed;
+			ref->cfg.pupd = pull;
+			ref->cfg.od = odata;
+			ref->cfg.af = alternate;
+#else
 			ref->active_cfg.mode = mode;
 			ref->active_cfg.otype = opendrain ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:360]: I on't change this as this code will soon be removed.

File: core/drivers/stm32_gpio.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:796]: i'll add an inline comment

**Comment by etienne-lms** [comment]
for info, CI checkpatch error reports are false positives (build error trace messages dumped in commit messages).

**Comment by etienne-lms** [comment]
review comments addressed

**Comment by etienne-lms** [comment]
may I squash the fixup commits?

File: core/arch/arm/plat-stm32mp1/shared_resources.c [hunk 10]
```diff
@@ -371,54 +371,66 @@ void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin)
...
 	unsigned int *pin = NULL;
 	size_t count = 0;
 	size_t n = 0;
 
-	stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
-	if (!count)
-		return;
+	res = stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
+	if (res || !count)
+		goto out;
```
Comment (GseoC) [inline core/arch/arm/plat-stm32mp1/shared_resources.c:384]: might as well return TEE_SUCCESS here as it precedes the allocation.

File: core/arch/arm/plat-stm32mp1/shared_resources.c [hunk 11]
```diff
@@ -371,54 +371,66 @@ void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin)
...
```
Comment (GseoC) [inline core/arch/arm/plat-stm32mp1/shared_resources.c:415]: ditto

**Comment by GseoC** [comment]
> may I squash the fixup commits?

For clearer view, yes please

**Comment by etienne-lms** [comment]
Fixup commits squashed + addressed comment on `stm32mp_register_secure_pinctrl()`/`stm32mp_register_non_secure_pinctrl()`.

**Comment by GseoC** [comment]
Thank you,

`Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

**Comment by etienne-lms** [comment]
tag applied
thanks

**Comment by etienne-lms** [comment]
I need to rebase due to https://github.com/OP-TEE/optee_os/pull/6098 being recently merged.

**Comment by etienne-lms** [comment]
@GseoC, I added a fixup commit to "drivers: stm32_gpio: support CFG_DRIVERS_PINCTRL", to address the rebase issue.
Could you please confirm (or infirm) your A-b tag with this change?

**Comment by jforissier** [comment]
I have restarted IBART which failed on a socket test (rcu_sched stalls), looks unrelated to this PR.

**Comment by etienne-lms** [comment]
thanks

**Comment by GseoC** [comment]
Seems fine, tag applies

**Comment by etienne-lms** [comment]
Some changes in the series, to finalize bumping _stm32_gpio.c_ to GPIO and PINCTRL frameworks:
- 1st commit is unmodified regarding previous review. @GseoC review tag is already applied to the commit.
"drivers: stm32_gpio: support CFG_DRIVERS_PINCTRL"

- The 5 next commit were review but I changed added functions prototypes to return `void` instead of a useless `TEE_Result`. @GseoC, I removed your review tag from those commits:
"drivers: stm32_gpio: add helper function stm32_gpio_pinctrl_bank_pin()" 
"drivers: stm32_gpio: add helper function stm32_pinctrl_set_secure_cfg()"
"plat-stm32mp1: shared_resources: support CFG_DRIVERS_PINCTRL"
"drivers: stm32_uart: support CFG_DRIVERS_PINCTRL" 
"drivers: stm32_i2c: support CFG_DRIVERS_PINCTRL"

- I added 5 new commits in the series, to finalize the evolution: enforce `CFG_DRIVERS_PINCTRL=y` and remove deprecated functions:
"plat-stm32mp1: conf: enable CFG_DRIVERS_PINCTRL"
"drivers: stm32_gpio: remove cases when CFG_DRIVERS_PINCTRL is disabled"
"drivers: stm32_uart: remove cases when CFG_DRIVERS_PINCTRL is disabled" 
"drivers: stm32_i2c: remove cases when CFG_DRIVERS_PINCTRL is disabled"
"drivers: stm32_gpio: move definitions to source file"

**Comment by etienne-lms** [comment]
Changed title has the seire finalizes stm32mp1 upgrades to pinctrl framework.
Checkpatch failure on a typo in a commit. I'll the 2 fixed after the review.

File: core/drivers/stm32_gpio.c [hunk 12]
```diff
@@ -56,6 +57,61 @@
 
 #define DT_GPIO_BANK_NAME0	"GPIOA"
 
+#define GPIO_MODE_INPUT		0x0
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:60]: Ok., i'll fix but I wonder if it's really needed?

File: core/drivers/stm32_gpio.c [hunk 13]
```diff
@@ -56,6 +57,61 @@
...
+	uint16_t af:		4;
+};
+
+/*
+ * Descrption of a pin and its 2 states muxing
+ *
+ * @bank: GPIO bank identifier as assigned by the platform
+ * @pin: Pin number in the GPIO bank
+ * @cfg: Pin configuration
+ * @active_cfg: Configuration in active state
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:106]: right, thanks

File: core/drivers/stm32_gpio.c [hunk 14]
```diff
@@ -902,6 +902,29 @@ void stm32_gpio_pinctrl_bank_pin(struct pinctrl_state *pinctrl,
 	*count = pin_count;
 }
 
+void stm32_pinctrl_set_secure_cfg(struct pinctrl_state *pinctrl, bool secure)
+{
+	size_t conf_index = 0;
+
+	if (!pinctrl)
+		return;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:842]: if there is no pinctrl, there is no pin to configure so returning straight is fine, IMO.

**Comment by GseoC** [comment]
I do not catch why the code style check fails, I guess it did not restart

`Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

**Comment by jforissier** [comment]
> I do not catch why the code style check fails, I guess it did not restart

It did, but the way the code style check works might be a bit counter-intuitive :wink: It is actually run twice:
- once for each commit in the PR, and
- once on the global diff (all commits squashed together if you wish)
The rationale is, we want to check the code but also the commit descriptions, so each commit is checked. When fixes are done as fixup commits (as opposed to force-pushing with corrected commits), the initial failure remain visible. However thanks to the second phase (called `checkdiff` in the log), you can be confident that the code style issues will be gone when squashing the fixups.

**Comment by GseoC** [comment]
> > I do not catch why the code style check fails, I guess it did not restart
> 
> It did, but the way the code style check works might be a bit counter-intuitive wink It is actually run twice:
> 
> * once for each commit in the PR, and
> * once on the global diff (all commits squashed together if you wish)
> 
> The rationale is, we want to check the code but also the commit descriptions, so each commit is checked (the `checkpatch` lines in the log). When fixes are done as fixup commits (as opposed to force-pushing with corrected commits), the initial failure remain visible. However thanks to the second phase (called `checkdiff` in the log), you can be confident that the code style issues will be gone when squashing the fixups.

Thanks for the explanation :)

**Comment by etienne-lms** [comment]
I'll sqaush the fixup commits. Once done, checkpatch should be happy.

**Comment by etienne-lms** [comment]
Thakns for the review.
Fixup commits squashed.
Review tag applied.

I also removed a suprious inline comment in comment "drivers: stm32_i2c: remove cases when CFG_DRIVERS_PINCTRL is disabled" (below). @GseoC, I preserved your review tag on that commit, I think it still applies :)
```patch
@@ -113,7 +113,6 @@ struct i2c_cfg {
  * @sec_cfg: I2C registers configuration storage
  * @pinctrl: Pin control configuration for the I2C bus in active state
  * @pinctrl_sleep: Pin control configuration for the I2C bus in standby state
- * @pinctrl_count: Number of PINCTRLs elements
  */
 struct i2c_handle_s {
        struct io_pa_va base;

**Pull request merged**